jdaviz/cli.py

```diff
@@ -53,3 +53,11 @@ def main(filename, layout='default'):
         sys.exit(Voila().launch_instance(argv=[]))
     finally:
         os.chdir(start_dir)
+
+if __name__ == '__main__':
+    import sys
+
+    if sys.platform == 'win32':
+        import asyncio
+        asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
+    main()
```

Comment:
This makes sense, but @nmearl, can you clarify why it *shouldn't* be called directly?  (after all, "CLI" seems like something one should be allowed to call on the command line)  And once we have the answer to that, @dtn5ah should put add a docstring to this file that explains the connection between these things!

Reply:
Definitely, I had assumed cli.main was supposed to be *an* entrypoint. Regardless, any entrypoint that leverages voila needs to have that asyncio line in order to have windows support until [#583](https://github.com/voila-dashboards/voila/issues/583) is fixed. So if it isn't here, then let me know where to move it!

Reply:
I think it's that the `jdaviz` script uses the `entry_points` mechanism in the setup file.  But I think it's still reasonable to allow direct calling the `cli.py` a la this approach, since the `entry_points` mechanism is pretty hairy to use any way other than as an actual command-line script.

But now that we're talking about it, if that's right, definitely should put a comment here along the lines of:
```suggestion
if __name__ == '__main__':
    # note that generally this CLI is triggered via the entry points mechanism, not calling this script directly
    import sys

    if sys.platform == 'win32':
        import asyncio
        asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
    main()
```
?

Also,  now that I think about it, is there a reason why this isn't at the top of `main()` instead of here?

Reply:
> Also, now that I think about it, is there a reason why this isn't at the top of main() instead of here?

Right, it's unclear why this is added for when the `cli.py` file is run explicitly, because this code chunk should not get run when done through the `entry_points` functionality. 

This should instead be placed in `main` or at the top of the file, unless you're not seeing the error when run using `jdaviz /file/path/ ...`?

Reply:
> can you clarify why it shouldn't be called directly?

Bah, I realize the context is weird. I was meant in reference to what how we document the usage of jdaviz to users. Perhaps a better phrasing would be "If a user were following instructions, they won't be running this file directly, so having things that only get run when the file is treated as the source file by the interpreter doesn't make sense."

Reply:
Yep I see your point, totally fair to move the asyncio check to the top of main.

I'm going to push back against relying solely on entry_points. `Jdaviz` is, by definition, a module that you can "launch" and thus I think there should be some way to execute our `Jdaviz` module as a script. I could see taking a step back and making sure we follow [PEP 338](https://www.python.org/dev/peps/pep-0338/) correctly, and actually create a proper `__main__()` for `Jdaviz`. But the concept of `runpy.run_module(modulename)` is a standard in the python community at this point, and I feel it makes sense that executing `Jdaviz` this way makes sense. At least, and this is how I dug up all of this, my vscode debugger requires either a specific script to run or a PEP 338 compliant module in order to run it. Right now if I try to hook into Jdaviz with my debugger, it just falls right through and doesn't execute anything because there's no `__main__()`, nor is there a `if __name__ == '__main__'` check. Implementing the former would be more prudent, but the latter check at least makes it functional with less involved architecting

Reply:
No need to push back. The fact that you use it this way is enough to justify letting `cli` run in the main scope. Though I'm not sure what you mean with your suggestion of following PEP 338, which deals with running top-level python modules existing on the system path as scripts, or your suggestion of having a `__main__()` function which was a rejected proposal back in 2002?

Regardless, it absolutely does not hurt us any to have `cli` be runnable independently of the entry points. My concern is that there shouldn't be content inside the `if` statement that won't get run when the application is run via the entry point (which should be solved by moving the event loop policy line it to the `main` function).

Fair point about debugging! I've always just had a non-tracked `debug.py` file in my package directories that I use for it, but I can see the use of being able to debug from a pre-existing file.

Reply:
Also as a follow up, I totally see your original point that `Jdaviz` was never supposed to exist, and that each of the tools were supposed to be separate. Yet somehow here we are, and so long as `Jdaviz` lives, we should probably have some kind of module level execution